### PR TITLE
8292657: Calling GetLocalXXX from virtual thread with thread parameter set to NULL returns carrier locals

### PR DESCRIPTION
**Pull request by sspitsyn**
If JVM TI GetLocalXXX/SetLocalXXX is called from a virtual thread with the thread parameter set to NULL (meaning current thread) then it should get/set the value of the locals in the virtual thread frames. Instead, it reads the carrier thread locals and/or crashes.

The fix is that the relevant checking of the jthread parameter for NULL and adjusting it to current thread is added.
It is done in new utility `function current_thread_obj_or_resolve_external_guard(jthread thread)`. For more convenient testing the same adjustment is done in the JVM TI extension function `GetCarrierThread()`.

The test serviceability/jvmti/vthread/GetSetLocalTest is updated to add previously missed test coverage.

The test serviceability/jvmti/vthread/VThreadTest has been updated to adopt to update behavior of the `GetCarrierThread`.

The fix was verified with the test/hotspot/jtreg/serviceability/jvmti/vthread/ tests.

The fix was also tested with the existing JVM TI and JDI tests to make sure no regressions are introduced.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8292657](https://bugs.openjdk.org/browse/JDK-8292657): Calling GetLocalXXX from virtual thread with thread parameter set to NULL returns carrier locals

### Reviewers
 * [Chris Plummer](https://openjdk.org/census#cjplummer) (@plummercj - **Reviewer**)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10051/head:pull/10051` \
`$ git checkout pull/10051`

Update a local copy of the PR: \
`$ git checkout pull/10051` \
`$ git pull https://git.openjdk.org/jdk pull/10051/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10051`

View PR using the GUI difftool: \
`$ git pr show -t 10051`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10051.diff">https://git.openjdk.org/jdk/pull/10051.diff</a>

</details>

### CODE REVIEW

File: test/hotspot/jtreg/serviceability/jvmti/vthread/GetSetLocalTest/libGetSetLocalTest.cpp [hunk 1]
```diff
@@ -312,6 +312,8 @@ static void
 test_GetSetLocal(jvmtiEnv *jvmti, JNIEnv* jni, jthread vthread, int depth, int frame_count, bool at_event) {
   Values values0 = { NULL, NULL, 1, 2L, (jfloat)3.2F, (jdouble)4.500000047683716 };
   Values values1 = { NULL, NULL, 2, 3L, (jfloat)4.2F, (jdouble)5.500000047683716 };
+  // jthread cthread = at_event ? get_carrier_thread(jvmti, jni, get_current_thread(jvmti, jni))
+  //                            : get_carrier_thread(jvmti, jni, vthread);
```
Comment (plummercj) [inline test/hotspot/jtreg/serviceability/jvmti/vthread/GetSetLocalTest/libGetSetLocalTest.cpp:316]: Is there a reason to keep this around?

Comment (sspitsyn) [reply test/hotspot/jtreg/serviceability/jvmti/vthread/GetSetLocalTest/libGetSetLocalTest.cpp:316]: You are right. Will remove this.

**Review by plummercj — APPROVED** [review]
Looks good.

**Comment by sspitsyn** [comment]
Thank you for review, Chris!

**Comment by sspitsyn** [comment]
Thank you for review, Alan!

**Pull request closed without merging**